dali/aux/optical_flow/turing_of/optical_flow_turing.cc

```diff
@@ -0,0 +1,169 @@
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+#include <dlfcn.h>
+#include "dali/aux/optical_flow/turing_of/optical_flow_turing.h"
+
+
+namespace dali {
+namespace optical_flow {
+
+namespace {
+
+void VerifySupport(NV_OF_STATUS status) {
+  switch (status) {
+    case NV_OF_SUCCESS:
+      return;
+    case NV_OF_ERR_OF_NOT_AVAILABLE:
+    case NV_OF_ERR_UNSUPPORTED_DEVICE:
+      throw unsupported_exception();
+    default:
+      DALI_FAIL("Optical flow failed, code: " + std::to_string(status));
+  }
+}
+
+}  // namespace
+
+OpticalFlowTuring::OpticalFlowTuring(dali::optical_flow::OpticalFlowParams params, size_t width,
+                                     size_t height, size_t channels) :
+        OpticalFlowAdapter<kernels::ComputeGPU>(params), width_(width), height_(height),
+        channels_(channels), device_(), context_(), stream_(0, true) {
+  DALI_ENFORCE(channels_ == 1 || channels_ == 3 || channels_ == 4);
+  DALI_ENFORCE(cuInitChecked(), "Failed to initialize driver");
+
+  int device_id;
+  CUDA_CALL(cudaGetDevice(&device_id));
+  CUDA_CALL(cuDeviceGet(&device_, device_id));
+  LoadTuringOpticalFlow("libnvidia-opticalflow.so");
+
+  SetInitParams(params);
+
+  context_ = CUContext(device_);
+  CUcontext ctx = context_;
+
+  TURING_OF_API_CALL(turing_of_.nvCreateOpticalFlowCuda(ctx, &of_handle_));
+  TURING_OF_API_CALL(turing_of_.nvOFSetIOCudaStreams(of_handle_, stream_, stream_));
+  VerifySupport(turing_of_.nvOFInit(of_handle_, &of_params_));
+
+  inbuf_.reset(
+          new OpticalFlowBuffer(of_handle_, width_, height_, turing_of_, NV_OF_BUFFER_USAGE_INPUT,
+                                NV_OF_BUFFER_FORMAT_ABGR8));
+  refbuf_.reset(
+          new OpticalFlowBuffer(of_handle_, width_, height_, turing_of_, NV_OF_BUFFER_USAGE_INPUT,
+                                NV_OF_BUFFER_FORMAT_ABGR8));
+  outbuf_.reset(
+          new OpticalFlowBuffer(of_handle_, (width_ + 3) / 4, (height_ + 3) / 4, turing_of_,
+                                NV_OF_BUFFER_USAGE_OUTPUT, NV_OF_BUFFER_FORMAT_SHORT2));
+}
+
+
+OpticalFlowTuring::~OpticalFlowTuring() {
+  inbuf_.reset(nullptr);
+  refbuf_.reset(nullptr);
+  outbuf_.reset(nullptr);
+  TURING_OF_API_CALL(turing_of_.nvOFDestroy(of_handle_));
+}
+
+
+void OpticalFlowTuring::CalcOpticalFlow(
+        dali::kernels::TensorView<dali::kernels::StorageGPU, const uint8_t, 3> reference_image,
+        dali::kernels::TensorView<dali::kernels::StorageGPU, const uint8_t, 3> input_image,
+        dali::kernels::TensorView<dali::kernels::StorageGPU, float, 3> output_image,
+        dali::kernels::TensorView<dali::kernels::StorageGPU, const float, 3> external_hints) {
```

Comment:
Are `external_hints` ignored?

Reply:
At this stage yes, it'll be there in next PR